### PR TITLE
Send code explanation within the message that gets sent to m…

### PR DESCRIPTION
**Pull request by baristaGeek**
## Description
When Run Watermelon is clicked, we also provide the code explanation in plain english on top the PR Body. It's language agnostic but it works best with the mainstream languages. 

Screen Shot 2022-05-04 at 11 45 15 AM

## Type of change
- New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -107,6 +107,7 @@
     "@vscode/test-electron": "^2.1.2",
     "axios": "^0.26.1",
     "glob": "^7.2.0",
-    "mocha": "^9.2.2"
+    "mocha": "^9.2.2",
+    "openai": "^2.0.5"
```
Comment (EstebanDalelR) [inline package.json:111]: why do we need this?

File: media/main.js [hunk 2]
```diff
@@ -50,6 +50,9 @@ $(document).ready(function () {
     $("#ghHolder").append(
       "<button class='run-watermelon'>Run Watermelon</button><br/>"
     );
+    $("#ghHolder").append(
+      `<p'>${explanation}</p>`
```
Comment (EstebanDalelR) [inline media/main.js:54]: Where is this explanation defined?

**Review by EstebanDalelR — CHANGES_REQUESTED** [review]

File: src/extension.ts [hunk 3]
```diff
@@ -102,6 +107,15 @@ export async function activate(context: vscode.ExtensionContext) {
   octokit = await credentials.getOctokit();
 
   vscode.window.onDidChangeTextEditorSelection(async (selection) => {
+    // // Get text of selected piece of code
+    let selectedCode = "";
+    if (selection.selections.length > 0) {
+      let selectedText = selection;
+      selectedCode= selectedText.textEditor.document.getText(selectedText.selections[0]);
```
Comment (EstebanDalelR) [inline src/extension.ts:114]: Interesting, my guess is that this would only select one line of code as only one param is passed but I might be wrong.
https://code.visualstudio.com/api/references/vscode-api#TextDocument

File: src/extension.ts [hunk 4]
```diff
@@ -138,6 +152,7 @@ class watermelonSidebar implements vscode.WebviewViewProvider {
   public static readonly viewType = "watermelon.sidebar";
 
   private _view?: vscode.WebviewView;
+  codeExplanation: any;
```
Comment (EstebanDalelR) [inline src/extension.ts:155]: Why do we need this here?

File: media/main.js [hunk 5]
```diff
@@ -209,6 +212,9 @@ $(document).ready(function () {
       case "author":
         authorName = message.author;
         break;
+      case "explanation":
+        console.log("message.explanation - main.js: ", message.explanation);
+        explanation = message.explanation;
```
Comment (EstebanDalelR) [inline media/main.js:217]: Where is this defined?

File: package.json (hunk 1 above)
Comment (baristaGeek) [reply package.json:111]: Open ai for obvious reasons. I don't know why mocha has a line diff at all, lol

File: src/extension.ts (hunk 3 above)
Comment (baristaGeek) [reply src/extension.ts:114]: It selects the whole block

File: media/main.js (hunk 5 above)
Comment (baristaGeek) [reply media/main.js]: Bad understanding of the code from my part. Moved stuff around. We're almost there.

File: src/extension.ts (hunk 4 above)
Comment (baristaGeek) [reply src/extension.ts]: Removed

**Comment by EstebanDalelR** [comment]
After a quick investigation, it is all working.
image

image

**Comment by EstebanDalelR** [comment]
On that image:
This was what was selected at the time:
`rl+shift+tab"
  ],
  "tab:jump:prefix": "command",
  "pane:next": "command+]",
  "pane:prev": "command+[",
  "pane:splitRight": "command+d",
  "pane:splitDown": "command+shift+d",
  "pane:close": "command+w",
  "editor:undo": "command+z",
  "editor:redo": "command+y",
  "editor:cut": "command+x",
  "editor:copy": "command+c",
  "editor:paste": "command+v",
  "editor:selectAll": "command+a",
  "editor:search": "command+f",
  "editor:search-close": "esc",
  "editor:movePreviousWord": "alt+left",
  "editor:moveNextWord": "alt+right",`

**Comment by EstebanDalelR** [comment]
I had the cursor on line 19, on index.html. It says it's Java? What `hello world`?
image

**Comment by baristaGeek** [comment]
> I had the cursor on line 19, on index.html. It says it's Java? What `hello world`? image

Happened the same to me and for a moment I thought it didn't understand CSS well. I latter selected the whole CSS selector and it worked fine. Should we consider this an edge case? Or simply force the user to run for various lines?

**Comment by EstebanDalelR** [comment]
> > I had the cursor on line 19, on index.html. It says it's Java? What `hello world`? image
> 
> Happened the same to me and for a moment I thought it didn't understand CSS well. I latter selected the whole CSS selector and it worked fine. Should we consider this an edge case? Or simply force the user to run for various lines?

I was under the impression that it selected a whole code block
https://github.com/watermelontools/wm-extension/pull/187#discussion_r864355600

**Review by EstebanDalelR — APPROVED** [review]

File: src/extension.ts [hunk 6]
```diff
@@ -102,6 +107,15 @@ export async function activate(context: vscode.ExtensionContext) {
...
 
   vscode.window.onDidChangeTextEditorSelection(async (selection) => {
+    // // Get text of selected piece of code
+    let selectedCode = "";
+    if (selection.selections.length > 0) {
+      let selectedText = selection;
+      selectedCode= selectedText.textEditor.document.getText(selectedText.selections[0]);
+    }
+    // Replace newlines with \n
+    selectedBlockOfCode = selectedCode.replace(/(\r\n|\n|\r)/gm,"");
```
[the remark this answers is not in the document]
Comment (baristaGeek) [reply src/extension.ts:117]: Definitely, if we decide to turn this experiment into a production-ready feature. Thanks for pointing this out

File: package.json (hunk 1 above)
Comment (baristaGeek) [reply package.json]: As discussed via Zoom, it was my bad. It's removed on the latest branch version anyway

File: src/extension.ts (hunk 3 above)
Comment (baristaGeek) [reply src/extension.ts:114]: What do you mean by fold exactly?

Comment (EstebanDalelR) [reply src/extension.ts:114]: It was my understanding it got the full code block, a function, a class, a switch statement, or something like that, but I tested and it grabs the _selection_.

File: src/extension.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (EstebanDalelR) [reply src/extension.ts:117]: So how is this not production?
I personally saw it on a user.
He uninstalled the app and his VSCode nav sped up immediately.

File: src/extension.ts (hunk 3 above)
Comment (baristaGeek) [reply src/extension.ts:114]: Yeah, that's the idea. Sorry if I wasn't clear when I described it with words.

File: src/extension.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (baristaGeek) [reply src/extension.ts:117]: What I mean is that if our user interviews show us good results we'll improve this and other aspects of the feature. 

Just the same way Paper needs a ton of improvement, but it's an experiment at the moment. Same way we ditched Slack button.

[the remark this answers is not in the document]
Comment (EstebanDalelR) [reply src/extension.ts:117]: Ok.
